### PR TITLE
Firewallns

### PR DESCRIPTION
**Pull request by NaveenSidhartha**
Added firewallns primitive with build, scrub and read verb functions.

### CODE REVIEW

File: firewallns.py [hunk 1]
```diff
@@ -0,0 +1,800 @@
...
+        table: str,
+        priority: int,
+        config_file=None,
+        rules=None,
+        nats=None,
+) -> Tuple[bool, str]:
+    """
+    description: |
+        1. Creates a /tmp/firewallns_<namespace>_<table>.conf file with nftable table config
+        2. Validates the nft file `sudo nft -c -f /tmp/firewallns_<namespace>_<table>.conf`
```
Comment (maria-walsh) [inline firewallns.py:95]: `sudo` should not be required in any payload request. The user sending the request via rcc.comms_ssh() should have the permissions to execute the payload.

File: firewallns.py [hunk 2]
```diff
@@ -0,0 +1,800 @@
...
+
+    # Block 09: Remove the temp nftables.conf file on Enabled PodNet
+    try:
+        remove_nftables_file, stdout, stderr = comms_ssh(
+            host_ip=enabled,
+            payload=payload_remvoe_nftables_file,
+            username='robot',
+        )
+    except CouldNotConnectException:
+        pass
```
Comment (maria-walsh) [inline firewallns.py:493]: Why is it not being reported that the rcc.comms_ssh() could not connect to the the enabled PodNet?

File: directorymain.py [hunk 3]
```diff
@@ -258,9 +265,10 @@ def read(
         return False, messages[3031]
 
     if exit_code != SUCCESS_CODE:
-        return False, f'{messages[3032]}\nExit Code: {exit_code}\nSTDOUT: {stdout}\nSTDERR: {stderr}'
+        return False, f'{messages[3032]}\nExit Code: {exit_code}\nSTDOUT: {disabled_stdout}\nSTDERR: {disabled_stderr}'
 
-    return True, messages[1000]
+    return True, f'{messages[1000]}. \nSTDOUT from Enabled PodNet: {enabled_stdout}' \
```
[the remark this answers is not in the document]
Comment (NaveenSidhartha) [reply directorymain.py:270]: This is for `read`, so primitive should give the output right?

File: firewallns.py (hunk 2 above)
Comment (NaveenSidhartha) [reply firewallns.py:493]: This step ie removing the temp /tmp/firewallns_<namespace>_<table>.conf is not compulsory so if it fails to connect is not a matter. So ignoring if fails to connect. But I think if a step is not compulsory then we should not use and ignore it if fails. So lets remove this step completely.

File: firewallns.py [hunk 4]
```diff
@@ -434,135 +434,195 @@ def build(
...
+        payload=payload_create_nftables_file,
+        username='robot',
+    )
+    if response['channel_code'] != 200:
+        msg = f'{messages[3050]}\n ' \
+              f'channel_code: {response["channel_code"]}s.\n' \
+              f'channel_message: {response["channel_message"]}\n' \
+              f'channel_error: {response["channel_error"]}'
+        return False, msg
+    if response['payload_code'] != 0:
```
Comment (maria-walsh) [inline firewallns.py:448]:
```suggestion
    if response['payload_code'] != SUCCESS_CODE:
```

As `SUCCESS_CODE` is a global variable, we can continue to use it for the `payload_code`

File: firewallns.py [hunk 5]
```diff
@@ -434,135 +434,195 @@ def build(
...
+        payload=payload_validate_nftables_file,
+        username='robot',
+    )
+    if response['channel_code'] != 200:
+        msg = f'{messages[3060]}\n ' \
+              f'channel_code: {response["channel_code"]}s.\n' \
+              f'channel_message: {response["channel_message"]}\n' \
+              f'channel_error: {response["channel_error"]}'
+        return False, msg
+    if response['payload_code'] != 0:
```
Comment (maria-walsh) [inline firewallns.py:467]:
```suggestion
    if response['payload_code'] != SUCCESS_CODE:
```

File: firewallns.py [hunk 6]
```diff
@@ -434,135 +434,195 @@ def build(
...
+        payload=payload_flush_table,
+        username='robot',
+    )
+    if response['channel_code'] != 200:
+        msg = f'{messages[3070]}\n ' \
+              f'channel_code: {response["channel_code"]}s.\n' \
+              f'channel_message: {response["channel_message"]}\n' \
+              f'channel_error: {response["channel_error"]}'
+        return False, msg
+    if response['payload_code'] != 0:
```
Comment (maria-walsh) [inline firewallns.py:486]:
```suggestion
    if response['payload_code'] != SUCCESS_CODE:
```

File: firewallns.py [hunk 7]
```diff
@@ -434,135 +434,195 @@ def build(
...
+        payload=payload_apply_nftables_file,
+        username='robot',
+    )
+    if response['channel_code'] != 200:
+        msg = f'{messages[3080]}\n ' \
+              f'channel_code: {response["channel_code"]}s.\n' \
+              f'channel_message: {response["channel_message"]}\n' \
+              f'channel_error: {response["channel_error"]}'
+        return False, msg
+    if response['payload_code'] != 0:
```
Comment (maria-walsh) [inline firewallns.py:505]:
```suggestion
    if response['payload_code'] != SUCCESS_CODE:
```

File: firewallns.py [hunk 8]
```diff
@@ -434,135 +434,195 @@ def build(
...
+        payload=payload_remvoe_nftables_file,
+        username='robot',
+    )
+    if response['channel_code'] != 200:
+        msg = f'{messages[3090]}\n ' \
+              f'channel_code: {response["channel_code"]}s.\n' \
+              f'channel_message: {response["channel_message"]}\n' \
+              f'channel_error: {response["channel_error"]}'
+        return False, msg
+    if response['payload_code'] != 0:
```
Comment (maria-walsh) [inline firewallns.py:524]:
```suggestion
    if response['payload_code'] != SUCCESS_CODE:
```

File: firewallns.py [hunk 9]
```diff
@@ -434,135 +434,195 @@ def build(
...
+        payload=payload_create_nftables_file,
+        username='robot',
+    )
+    if response['channel_code'] != 200:
+        msg = f'{messages[3100]}\n ' \
+              f'channel_code: {response["channel_code"]}s.\n' \
+              f'channel_message: {response["channel_message"]}\n' \
+              f'channel_error: {response["channel_error"]}'
+        return False, msg
+    if response['payload_code'] != 0:
```
Comment (maria-walsh) [inline firewallns.py:544]:
```suggestion
    if response['payload_code'] != SUCCESS_CODE:
```

File: firewallns.py [hunk 10]
```diff
@@ -434,135 +434,195 @@ def build(
...
+        payload=payload_validate_nftables_file,
+        username='robot',
+    )
+    if response['channel_code'] != 200:
+        msg = f'{messages[3110]}\n ' \
+              f'channel_code: {response["channel_code"]}s.\n' \
+              f'channel_message: {response["channel_message"]}\n' \
+              f'channel_error: {response["channel_error"]}'
+        return False, msg
+    if response['payload_code'] != 0:
```
Comment (maria-walsh) [inline firewallns.py:563]:
```suggestion
    if response['payload_code'] != SUCCESS_CODE:
```

File: firewallns.py [hunk 11]
```diff
@@ -434,135 +434,195 @@ def build(
...
+        payload=payload_flush_table,
+        username='robot',
+    )
+    if response['channel_code'] != 200:
+        msg = f'{messages[3120]}\n ' \
+              f'channel_code: {response["channel_code"]}s.\n' \
+              f'channel_message: {response["channel_message"]}\n' \
+              f'channel_error: {response["channel_error"]}'
+        return False, msg
+    if response['payload_code'] != 0:
```
Comment (maria-walsh) [inline firewallns.py:582]:
```suggestion
    if response['payload_code'] != SUCCESS_CODE:
```

File: firewallns.py [hunk 12]
```diff
@@ -434,135 +434,195 @@ def build(
...
+        payload=payload_apply_nftables_file,
+        username='robot',
+    )
+    if response['channel_code'] != 200:
+        msg = f'{messages[3130]}\n ' \
+              f'channel_code: {response["channel_code"]}s.\n' \
+              f'channel_message: {response["channel_message"]}\n' \
+              f'channel_error: {response["channel_error"]}'
+        return False, msg
+    if response['payload_code'] != 0:
```
Comment (maria-walsh) [inline firewallns.py:601]:
```suggestion
    if response['payload_code'] != SUCCESS_CODE:
```

File: firewallns.py [hunk 13]
```diff
@@ -434,135 +434,195 @@ def build(
...
+        payload=payload_remvoe_nftables_file,
+        username='robot',
+    )
+    if response['channel_code'] != 200:
+        msg = f'{messages[3140]}\n ' \
+              f'channel_code: {response["channel_code"]}s.\n' \
+              f'channel_message: {response["channel_message"]}\n' \
+              f'channel_error: {response["channel_error"]}'
+        return False, msg
+    if response['payload_code'] != 0:
```
Comment (maria-walsh) [inline firewallns.py:620]:
```suggestion
    if response['payload_code'] != SUCCESS_CODE:
```

File: firewallns.py [hunk 14]
```diff
@@ -660,30 +720,42 @@ def scrub(
...
+        payload=payload_flush_table,
+        username='robot',
+    )
+    if response['channel_code'] != 200:
+        msg = f'{messages[3020]}\n ' \
+              f'channel_code: {response["channel_code"]}s.\n' \
+              f'channel_message: {response["channel_message"]}\n' \
+              f'channel_error: {response["channel_error"]}'
+        return False, msg
+    if response['payload_code'] != 0:
```
Comment (maria-walsh) [inline firewallns.py:734]:
```suggestion
    if response['payload_code'] != SUCCESS_CODE:
```

File: firewallns.py [hunk 15]
```diff
@@ -660,30 +720,42 @@ def scrub(
...
+        payload=payload_flush_table,
+        username='robot',
+    )
+    if response['channel_code'] != 200:
+        msg = f'{messages[3030]}\n ' \
+              f'channel_code: {response["channel_code"]}s.\n' \
+              f'channel_message: {response["channel_message"]}\n' \
+              f'channel_error: {response["channel_error"]}'
+        return False, msg
+    if response['payload_code'] != 0:
```
Comment (maria-walsh) [inline firewallns.py:753]:
```suggestion
    if response['payload_code'] != SUCCESS_CODE:
```

File: firewallns.py [hunk 16]
```diff
@@ -817,57 +890,57 @@ def read(
...
+              f'channel_message: {response["channel_message"]}\n' \
+              f'channel_error: {response["channel_error"]}'
+        message_list.append(msg)
         success = False
-        message_list.append(messages[3020])
-        exit_code = None
-        stdout = None
-
-    if exit_code is None and exit_code != SUCCESS_CODE:
+    if response['payload_code'] != 0:
```
Comment (maria-walsh) [inline firewallns.py:913]:
```suggestion
    if response['payload_code'] != SUCCESS_CODE:
```

File: firewallns.py [hunk 17]
```diff
@@ -817,57 +890,57 @@ def read(
...
+              f'channel_message: {response["channel_message"]}\n' \
+              f'channel_error: {response["channel_error"]}'
+        message_list.append(msg)
         success = False
-        message_list.append(messages[3030])
-        exit_code = None
-        stdout = None
-
-    if exit_code is None and exit_code != SUCCESS_CODE:
+    if response['payload_code'] != 0:
```
Comment (maria-walsh) [inline firewallns.py:936]:
```suggestion
    if response['payload_code'] != SUCCESS_CODE:
```

File: firewallns.py [hunk 18]
```diff
@@ -817,57 +890,57 @@ def read(
...
+        msg = f'{messages[3031]}\n ' \
+              f'payload_code: {response["payload_code"]}s.\n' \
+              f'payload_message: {response["payload_message"]}\n' \
+              f'payload_error: {response["payload_error"]}'
+        message_list.append(msg)
         success = False
-        message_list.append(messages[3031])
 
-    data_dict[enabled] = stdout
+    data_dict[enabled] = response['payload_message']
```
Comment (maria-walsh) [inline firewallns.py:944]:
```suggestion
    data_dict[disabled] = response['payload_message']
```

**Review by maria-walsh — CHANGES_REQUESTED** [review]

**Comment by maria-walsh** [comment]
@NaveenSidhartha Can you revert the changes made to the directorymain primitive? I have an PR open updating it to the change in the rcc.

Once this file is reverted, the PR can be merged

**Comment by NaveenSidhartha** [comment]
> @NaveenSidhartha Can you revert the changes made to the directorymain primitive? I have an PR open updating it to the change in the rcc.
> 
> Once this file is reverted, the PR can be merged

Done

**Review by maria-walsh — APPROVED** [review]

**Pull request merged**